src/operator/warp.cc

```diff
@@ -0,0 +1,277 @@
+  int grids_strideHeight = grids_strideWidth*grid.size(2);
+  int grids_strideBatch = grids_strideHeight*grid.size(1);
+
+  Dtype *inputImages_data, *output_data, *grids_data;
+  inputImages_data = data.dptr_;
+  output_data = out.dptr_;
+  grids_data = grid.dptr_;
+
+  int b, yOut, xOut;
+
+  for (b=0; b < batchsize; b++) {
+    for (yOut=0; yOut < output_height; yOut++) {
+      for (xOut=0; xOut < output_width; xOut++) {
+        // read the grid
+        double xf = grids_data[b*grids_strideBatch +
+         yOut*grids_strideHeight + xOut*grids_strideWidth];
+        double yf = grids_data[b*grids_strideBatch +
+         yOut*grids_strideHeight + xOut*grids_strideWidth + 1];
+
+        // get the weights for interpolation
+        int yInTopLeft, xInTopLeft;
+        double yWeightTopLeft, xWeightTopLeft;
+
+        double xcoord = xf + xOut;
+        xInTopLeft = floor(xcoord);
+        xWeightTopLeft = std::max(1 - fabs(xcoord - xInTopLeft), 0.0);
+
+        double ycoord = yf + yOut;
+        yInTopLeft = floor(ycoord);
+        yWeightTopLeft = std::max(1 - fabs(ycoord - yInTopLeft), 0.0);
+
+        int outAddress = output_strideBatch * b + output_strideHeight *
+         yOut + output_strideWidth * xOut;
+        int inTopLeftAddress = inputImages_strideBatch * b +
+         inputImages_strideHeight * yInTopLeft + inputImages_strideWidth * xInTopLeft;
+        int inTopRightAddress = inTopLeftAddress + inputImages_strideWidth;
+        int inBottomLeftAddress = inTopLeftAddress + inputImages_strideHeight;
+        int inBottomRightAddress = inBottomLeftAddress + inputImages_strideWidth;
+
+        double v = 0;
+        double inTopLeft = 0;
+        double inTopRight = 0;
+        double inBottomLeft = 0;
+        double inBottomRight = 0;
+
+        // we are careful with the boundaries
+        bool topLeftIsIn = xInTopLeft >= 0 && xInTopLeft <= inputImages_width-1 &&
+         yInTopLeft >= 0 && yInTopLeft <= inputImages_height-1;
+        bool topRightIsIn = xInTopLeft+1 >= 0 && xInTopLeft+1 <= inputImages_width-1 &&
+         yInTopLeft >= 0 && yInTopLeft <= inputImages_height-1;
+        bool bottomLeftIsIn = xInTopLeft >= 0 && xInTopLeft <= inputImages_width-1 &&
+         yInTopLeft+1 >= 0 && yInTopLeft+1 <= inputImages_height-1;
+        bool bottomRightIsIn = xInTopLeft+1 >= 0 && xInTopLeft+1 <= inputImages_width-1 &&
+         yInTopLeft+1 >= 0 && yInTopLeft+1 <= inputImages_height-1;
+
+        int t;
+        // interpolation happens here
+        for (t=0; t < inputImages_channels; t++) {
+           if (topLeftIsIn) inTopLeft = inputImages_data[inTopLeftAddress + t];
+           if (topRightIsIn) inTopRight = inputImages_data[inTopRightAddress + t];
+           if (bottomLeftIsIn) inBottomLeft = inputImages_data[inBottomLeftAddress + t];
+           if (bottomRightIsIn) inBottomRight = inputImages_data[inBottomRightAddress + t];
+
+           v = xWeightTopLeft * yWeightTopLeft * inTopLeft

```

Comment:
this is not efficient. consider make calculate xWeightTopLeft * yWeightTopLeft inside line 92 if..else. when inTopLeft is zero, we don't need calculate xWeightTopLeft * yWeightTopLeft.